tempesta_fw/connection.h

```diff
@@ -54,20 +54,26 @@ enum {
  * Session/Presentation layer (in OSI terms) handling.
  *
  * @proto	- protocol handler. Base class, must be first;
- * @list	- list of connections with the @peer;
- * @msg_queue	- messages queue to be sent over the connection;
- * @msg		- currently processing (receiving) message;
+ * @list	- member in the list of connections with @peer;
+ * @msg_queue	- queue of messages to be sent over the connection;
+ * @msg_qlock	- lock for accessing @msg_queue;
+ * @refcnt	- number of users of the connection structure;
+ * @msg		- message that is currently being processed;
  * @peer	- TfwClient or TfwServer handler;
- * @sk		- appropriate sock handler;
+ * @sk		- an appropriate sock handler;
+ * @splock	- lock for accessing @peer and @sk;
  */
 typedef struct {
 	SsProto			proto;
 	struct list_head	list;
 	struct list_head	msg_queue;
+	spinlock_t		msg_qlock;
+	atomic_t		refcnt;
 
 	TfwMsg			*msg;
 	TfwPeer 		*peer;
 	struct sock		*sk;
+	spinlock_t		splock;
```

Comment:
It'd be better to place the member near from msg_qlock` to avoid unnecessary memory holes due to alignment.